From f2ff500163b88c4160f110b58ec84f0881cea1d5 Mon Sep 17 00:00:00 2001
From: John Hawthorn <john@hawthorn.email>
Date: Sun, 2 Jun 2024 23:39:36 -0700
Subject: [PATCH] Avoid unnecessary writes to ISEQ during GC

On mark we check whether a callcache has been invalidated and if it has
we replace it with the empty callcache, rb_vm_empty_cc(). However we
also consider the empty callcache to not be active, and so previously
would overwrite it with itself.

These additional writes are problematic because they may force
Copy-on-Write to occur on the memory page, increasing system memory use.

(cherry picked from commit 9d6b8806a4e0801c260f628e48a21ae94cb3fc91)

First commit of https://github.com/ruby/ruby/pull/10899, backported to
Ruby 3.3 for https://gitlab.com/gitlab-org/gitlab/-/work_items/634879.
---
 iseq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/iseq.c b/iseq.c
index d28d68206b..08e84ec551 100644
--- a/iseq.c
+++ b/iseq.c
@@ -346,7 +346,7 @@ rb_iseq_mark_and_move(rb_iseq_t *iseq, bool reference_updating)
                 if (cc_is_active(cds[i].cc, reference_updating)) {
                     rb_gc_mark_and_move_ptr(&cds[i].cc);
                 }
-                else {
+                else if (cds[i].cc != rb_vm_empty_cc()) {
                     cds[i].cc = rb_vm_empty_cc();
                 }
             }
-- 
2.52.0

